test/jdk/java/nio/jni/NewDirectByteBuffer.java

```diff
@@ -0,0 +1,136 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @bug 8299684
+ * @summary Unit test for the JNI function NewDirectByteBuffer
+ * @requires (sun.arch.data.model == "64" & os.maxMemory >= 8g)
+ * @modules java.base/jdk.internal.misc
+ * @run junit/othervm/native NewDirectByteBuffer
+ */
+public class NewDirectByteBuffer {
+    private static final Unsafe UNSAFE;
+    static {
+        System.loadLibrary("NewDirectByteBuffer");
+        UNSAFE = Unsafe.getUnsafe();
+    }
+
+    private static final void checkBuffer(ByteBuffer buf, long capacity) {
+        // Verify that the JNI function returns the correct capacity
+        assertEquals(capacity, getDirectByteBufferCapacity(buf),
+            "GetDirectBufferCapacity returned unexpected value");
+
+        // Verify that the initial state values are correct
+        assertTrue(buf.isDirect(), "Buffer is not direct");
+        assertFalse(buf.hasArray(), "Buffer has an array");
+        if (capacity > 0) {
+            assertTrue(buf.hasRemaining(), "Buffer has no remaining values");
+        }
+        assertFalse(buf.isReadOnly(), "Buffer s read-only");
+        assertEquals(capacity, buf.capacity(),
+            "Buffer::capacity returned unexpected value");
+        assertEquals(0L, buf.position(),
+            "Buffer::position returned unexpected value");
+        assertEquals(capacity, buf.limit(),
+            "Buffer::limit returned unexpected value");
+
+        // Verify that the various state mutators work correctly
+        int halfPos = buf.capacity()/2;
+        buf.position(halfPos);
+        assertEquals(halfPos, buf.position(),
+            "Position not set to halfPos");
+        assertEquals(buf.capacity() - halfPos, buf.remaining(),
+            "Remaining not capacity - halfPos");
+
+        buf.mark();
+
+        int twoThirdsPos = 2*(buf.capacity()/3);
+        buf.position(twoThirdsPos);
+        assertEquals(twoThirdsPos, buf.position(),
+            "Position not set to twoThirdsPos");
+        assertEquals(buf.capacity() - twoThirdsPos, buf.remaining(),
+            "Remaining != capacity - twoThirdsPos");
+
+        buf.reset();
+        assertEquals(halfPos, buf.position(),
+            "Buffer not reset to halfPos");
+
+        buf.limit(twoThirdsPos);
+        assertEquals(twoThirdsPos, buf.limit(),
+            "Limit not set to twoThirdsPos");
+        assertEquals(twoThirdsPos - halfPos, buf.remaining(),
+            "Remaining != twoThirdsPos - halfPos");
+
+        buf.position(twoThirdsPos);
+        assertFalse(buf.hasRemaining(), "Buffer has remaining values");
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {0L, 1L, (long)Integer.MAX_VALUE/2,
+        (long)Integer.MAX_VALUE - 1, (long)Integer.MAX_VALUE})
+    void legalCapacities(long capacity) {
+        long addr = UNSAFE.allocateMemory(capacity);
```

Comment:
I think David was asking for the test to continue when malloc fails. Now that the test is changed to use Unsafe.allocateMemory it means it has to catch OOME.

Reply:
Added catch of OOME in f9350695c3a6f90e3acdf2176d78c06406e49387.